hw/ip/kmac/doc/theory_of_operation.md

```diff
@@ -143,69 +143,27 @@ Default design parameters assume the system characteristics as below:
-
-In code, this looks something like:
-```c
-/**
- * Absorb input data into the Keccak computation.
- *
- * Assumes that the KMAC block is in the "absorb" state; it is the caller's
- * responsibility to check before calling.
- *
- * @param in Input buffer.
- * @param in_len Length of input buffer (bytes).
- * @return Number of bytes written.
- */
-size_t kmac_absorb(const uint8_t *in, size_t in_len) {
-    // Read FIFO depth from the status register.
-    uint32_t status = abs_mmio_read32(kBase + KMAC_STATUS_REG_OFFSET);
-    uint32_t fifo_depth =
-        bitfield_field32_read(status, KMAC_STATUS_FIFO_DEPTH_FIELD);
-
-    // Calculate the remaining space in the FIFO using auto-generated KMAC
-    // parameters and take the minimum of that space and the input length.
-    size_t free_entries = (KMAC_PARAM_NUM_ENTRIES_MSG_FIFO - fifo_depth);
-    size_t max_len = free_entries * KMAC_PARAM_NUM_BYTES_MSG_FIFO_ENTRY;
-    size_t write_len = (in_len < max_len) ? in_len : max_len;
-
-    // Note: this example uses byte-writes for simplicity, but in practice it
-    // would be more efficient to use word-writes for aligned full words and
-    // byte-writes only as needed at the beginning and end of the input.
-    for (size_t i = 0; i < write_len; i++) {
-      abs_mmio_write8(kBase + KMAC_MSG_FIFO_REG_OFFSET, in[i]);
-    }
-
-    return write_len;
-}
-```
-
-The method recommended above is always safe.
-However, in specific contexts, it may be okay to skip polling `STATUS.fifo_depth`.
-Normally, KMAC will process data faster than software can write it, and back pressure on the FIFO interface, should ensure that writes from software will simply block until KMAC can process messages.
-The only reason for polling, then, is to prevent a specific deadlock scenario:
-1. Software has configured KMAC to wait forever for entropy.
-2. There is a problem with the EDN, so entropy is never coming.
-3. The FIFO is full and KMAC is waiting for entropy to process it.
-
-If either the entropy wait timer is nonzero or `kmac_en` is false (so KMAC will not be refreshing entropy), it is safe to write to the FIFO without polling `STATUS.fifo_depth`.
-However, this should be done carefully, and tests should always cover the scenario in which EDN is locked up.
+If software pushes data to the Message FIFO while it is full, the write operation is blocked until there is again space in the FIFO meaning the processor is effectively stalled.
+In case the FIFO ran full because the SHA3 engine is currently running, the stalling doesn't take more than 100 clock cycles, meaning this mechanism conveniently prevents data loss and relieves software from polling the [`STATUS.fifo_depth`](registers.md#status--fifo_depth) field before writing data.
+
+However, the FIFO can also run full due to the KMAC block waiting for fresh entropy from EDN.
```

Comment:
Similarly, I think I'd tweak this. How about this?
```
The FIFO can also become full because the KMAC block is waiting for fresh entropy from EDN.
```